# Trabajo Práctico Integrador



## Urban Flow



## Integrantes
- Cristian Vera
- Darío Verdún
- Sergio Sánchez
- Eduardo Saldívia


## Sprint 1

## Consideraciones Generales e Importantes

- Desarrollo base: Los puntos solicitados son la base del trabajo práctico. Es posible la ampliación sin restricciones del mismo.

- Organización del Notebook:
  - Agregar antes del ejercicio 01 una celda donde se importan todas librerías necesarias.

  - Utilizar un título para cada celda y escribir cada funcionalidad en una celda con título. Tener en cuenta agrupar las funcionalidades afines.

## Versionado y Trabajo en Equipo

- La entrega puede entregarse en grupos de máximo 6 personas (mínimo 1 / máximo 6) y la defensa es individual. Los grupos deben mantenerse a lo largo del trabajo integrador.

- Se debe considerar que es un trabajo en equipo y que cualquiera de los integrantes puede realizar modificaciones para cada día de trabajo.

- Control de Versiones: Utilizar herramientas de versionado. Vincularlo por token a tu cuenta personal de versionado en la nube.

- Colaboradores: Brindar acceso al repositorio a las siguientes cuentas:

  - lcd-sa182@ugr.edu.ar

  - fpasinato@ugr.edu.ar

- Changelog: Cada ejercicio/punto del trabajo práctico es considerado como un día distinto de trabajo. Debe quedar constancia en un archivo CHANGELOG.md de los cambios generados cada día.

- Agregar una variable llamada desactivar_git_push que solamente me permita desactivar los "git push".  

- El archivo README.md debe contener el sprint en el que nos encontramos y debe contener:
  - El objetivo.
  - La introducción y el contexto del sprint de trabajo.


## Criterios de evaluación

- Funcionalidad:

  - El código debe ejecutar sin errores al ejecutar "Ejecutar todo" (Ctrl+F9).

  - Se debe implementar la totalidad de las funcionalidades solicitadas.
  
  - La lógica del programa debe ser correcta y consistente.
  
  - Utilización de librerías acorde a los problemas a resolver.

  - Recordar antes de entregar de validar el notebook con los siguientes pasos. Ir al menú Runtine/Ejecutar -> Restart session and run all/Reiniciar la sessión y ejecutar todo

- Claridad y organización:

  - El código debe estar bien organizado y dividido.

  - Utilizar diferentes celdas de código para resolver cada solicitud/punto dentro de un ejercicio específico con su comentario "title" correspondiente a la solución implementada.
  
  - Se deben utilizar nombres de variables que describan su contenido/función.
  
  - Se debe agregar comentarios explicativos y necesarios según estandares.

  - Todos los imports deben ubicarse en la primer celda de ejecución arriba de la descripción del "Ejercicio 01".

  - No se admite la sentencia "git add ." para simplificar el traceo.
  
  - No se admiten comentarios innecesarios dentro del código. Solo comentar funciones, clases, métodos y propiedades de clase.

- Estilo y presentación:

  - Se debe usar una indentación consistente y un formato de código adecuado.

  - Se deben evitar comentarios redundantes o irrelevantes.

  - Se debe presentar el código de forma ordenada y profesional.
  
  - Se debe utilizar [type hints](https://docs.python.org/3/library/typing.html).
  
  - Se deben utilizar comentarios según el [pep8](https://peps.python.org/pep-0008/).

- Herramientas:
  
  - Utilizar principalmente librerías vistas durante la cursada. No se admiten librerías similares a las vistas en la materia.

  - Utilizar librerías extras con criterios fundamentados, es decir, no utilizar por utilizar.

  - Indentar a 2 espacios y cortar a 80 caracteres según estandares.

  - No se puede utilizar gdrive.
  
  - No se pueden utilizar herramientas alternativas a las vistas durante la cursada.

## Condiciones de Entrega
- Formato de Archivo: Los estudiantes deberán enviar un archivo con extensión .ipynb.

- Nomenclatura: El nombre original del archivo debe reemplazar los valores indicados. Ejemplo: 0X_Price_Manager-Apellidos_Nombres.ipynb.

- Estado de la entrega:
  - Deberá entregarse ejecutado y con los datos precargados.
  - Deberá entregarse el desarrollo sobre una copia de la plantilla original.

- Canal de Entrega: La entrega se realiza en el aula virtual y solo se aceptan archivos (cualquier otra forma se considerará inválida). Solo se recibirán trabajos por correo adjunto a lcd-sa182@ugr.edu.ar si el plazo en la plataforma ha finalizado. No se aceptarán enlaces sueltos.


## Recomendaciones y Consecuencias del Incumplimiento
- Recomendaciones: Comenzar a trabajar con suficiente antelación. Leer y comprender el enunciado en su totalidad. Realizar consultas ante dudas al correo: lcd-sa182@ugr.edu.ar o fpasinato@ugr.edu.ar. La entrega en tiempo y forma es fundamental.

- Incumplimiento: Los trabajos que no cumplan con los requisitos no serán tomados en cuenta como nota final. Solo serán corregidos y evaluados con nota para la devolución correspondiente. Se considera incumplimiento a:

  - Entregas fuera de término.

  - Trabajos que no se entreguen sobre la plantilla de trabajo estipulada.

  - Trabajos que no se ejecuten completos (que contengan errores intermedios al dar "Ejecutar todo").

---

## Objetivo

El objetivo principal de este proyecto es aplicar los conocimientos adquiridos en para el versionado de código, la organización, limpieza del código y la utilización de pandas.


## Introducción y Contexto del problema


### Sprint 1

La localidad llamada Vaalserberg de Bélgica se encuentra en la zona fronteriza y limita con los paises de Países Bajos y Alemania. Esta localidad cuenta con un sistema de radares urbanos para la detección de infracciones por exceso de velocidad. Los registros históricos provienen de sistemas heredados, el cuál presenta errores de formato, faltante de datos generando registros inconsistentes en el nuevo sistema.

Debemos analizar y depurar los datos de viejo sistema para obtener información relevante sobre las infracciones y de está forma en el futuro poder incorporar los datos al nuevo sistema sin inconsistencias.

Descargar el dataset [speeding fines](https://raw.githubusercontent.com/HAD141/datasets/refs/heads/main/TrabajosPracticos/urban_flow/speeding_fines.csv) https://raw.githubusercontent.com/HAD141/datasets/refs/heads/main/TrabajosPracticos/urban_flow/speeding_fines.csv

El dataset contiene información histórica de multas por exceso de velocidad y presenta errores que deberán ser tratados para evitar inconsistencias.


---

# Importación de librerías

In [22]:
# Control de push
desactivar_git_push = True # Cambiar a False para hacer push.

In [ ]:
# ==========================================
# CONFIGURACIÓN PERSONAL
# Cada integrante debe modificar estos datos
# ==========================================

nombre_git = "Cristian Vera"
email_git = "TU_MAIL_DE_GITHUB"
rama_personal = "cristian" # cristian, dario, sergio ó eduardo
usuario_github = "ckriztian"



# ==========================================
# CLONAR REPOSITORIO
# Ejecutar al comenzar una nueva sesión
# ==========================================

!git clone https://github.com/ckriztian/lcd-tp-urban-flow.git
%cd /content/lcd-tp-urban-flow

# ==========================================
# CONFIGURAR IDENTIDAD EN GIT
# ==========================================

!git config user.name "{nombre_git}"
!git config user.email "{email_git}"
!git config --list | grep user

# ==========================================
# TRAER RAMAS REMOTAS
# ==========================================

!git fetch --all
!git branch -a


# ==========================================
# ENTRAR A LA RAMA PERSONAL Y ACTUALIZARLA
# ==========================================

!git checkout {rama_personal}
!git pull origin {rama_personal} --rebase

# ==========================================
# VERIFICAR ESTADO DEL REPOSITORIO
# ==========================================

%cd /content/lcd-tp-urban-flow
!git status

In [ ]:
import os
import requests
import pandas as pd

## Ejercicio 01

> Puntos: 1

Inicialización y configuración de la herramienta de versionado. Vamos a trabajar sobre la rama llamada "Sprint_1".

La estructura de directorios es:

```
├── urban_flow
│   ├── data
│   │   ├── interim (almacena dataset procesados de pasos intermedios)
│   │   │   ├── plots
│   │   ├── processed (almacena dataset procesados finales para entregar a otra aplicación)
│   │   ├── raw (almacena los dataset en crudo, sin procesar)
```


In [ ]:
#@title Inicialización y configuración de la herramienta de versionado y estructura de directorios
os.makedirs("urban_flow/data/raw", exist_ok=True)
os.makedirs("urban_flow/data/interim/plots", exist_ok=True)
os.makedirs("urban_flow/data/processed", exist_ok=True)

print("Estructura de directorios creada correctamente")

Estructura de directorios creada correctamente


## Ejercicio 02

> Puntos: 1

Los desarrollos que deben realizar son:

- Descargar el dataset raw original y almacenarlo en `urban_flow/data/raw`.

- Mostrar las 5 primeras filas.

- Analizar tipos de datos

- Contar los valores nulos.



In [ ]:

#@title Descargar el dataset raw original y almacenarlo en urban_flow/data/raw.
# URL del dataset
url: str = (
  "https://raw.githubusercontent.com/HAD141/datasets/"
  "refs/heads/main/TrabajosPracticos/urban_flow/speeding_fines.csv"
)

# Ruta destino (estructura ya creada en el repo)
file_path: str = "urban_flow/data/raw/speeding_fines.csv"

# Descarga del dataset
response = requests.get(url)

# Validación de descarga
if response.status_code != 200:
  raise Exception(
    f"Error al descargar el dataset. Código: {response.status_code}"
  )

# Guardado en disco
with open(file_path, "wb") as file:
  file.write(response.content)

# Carga del dataset en DataFrame
df: pd.DataFrame = pd.read_csv(file_path)

In [ ]:
#@title Visualización de las primeras 5 filas del dataset
df.head()

,multa_id,patente,fecha,hora,velocidad_registrada,velocidad_maxima,ubicacion,radar_id,estado_multa
0,1,FD 1475,24/09/2021,09:01 AM,NaN,60,Av. Siempre Viva,R01,APELADA
1,2,0695 DVH,19/05/2020,18:01,NaN,40,Av Siempre Viva,NaN,APELADA
2,3,WU 341 A,2022/07/17,09:01 AM,107.0,60,AV LIBERTADOR,R03,PAGADA
3,4,GN64 OTP,invalid_date,invalid_hour,NaN,60,Ruta 9,NaN,APELADA
4,5,BX 55 DOG,2021/01/21,invalid_hour,94.0,60,Av. Siempre Viva,NaN,ERROR


El dataset histórico de multas proviene de un sistema de radares urbanos implementado en la localidad fronteriza de Vaalserberg (Bélgica), donde se registran infracciones por exceso de velocidad en distintos puntos de control. Sin embargo, los datos corresponden a un sistema heredado que presenta múltiples inconsistencias, errores de formato y valores faltantes, lo que obliga a realizar un proceso de limpieza y normalización antes de su integración al nuevo sistema.

In [ ]:
#@title Analizar tipos de datos
df.dtypes


,0
multa_id,int64
patente,object
fecha,object
hora,object
velocidad_registrada,float64
velocidad_maxima,int64
ubicacion,object
radar_id,object
estado_multa,object


El dataset combina variables numéricas, categóricas y campos de texto. Se observan correctamente tipificadas las variables multa_id (int64), velocidad_registrada (float64) y velocidad_maxima (int64), mientras que la mayoría de las columnas aparecen como object.

Las variables fecha y hora están en formato texto, lo que evidencia inconsistencias y la necesidad de conversión a formatos temporales. En general, el dataset requiere estandarización de tipos para garantizar su correcto análisis y procesamiento posterior.


In [ ]:
#@title Contar los valores nulos.
df.isnull().sum()

,0
multa_id,0
patente,50
fecha,0
hora,0
velocidad_registrada,1957
velocidad_maxima,0
ubicacion,0
radar_id,1002
estado_multa,0


El análisis de valores nulos muestra que la mayoría de las variables no presentan datos faltantes, sin embargo existen columnas críticas con pérdidas significativas de información.

Las variables más afectadas son velocidad_registrada con 1957 valores nulos y radar_id con 1002 valores faltantes, lo que puede impactar directamente en el análisis de infracciones y la trazabilidad de los registros. En menor medida, patente presenta 50 valores nulos.

El resto de las columnas no presenta valores faltantes, lo que indica una buena consistencia parcial del dataset, aunque con problemas concentrados en variables clave de medición y identificación.

## Ejercicio 03

> Puntos: 2

Utilizando Pandas:

- Normalizar las de fechas con el formato 'YYYY-MM-DD', en caso de fechas inválidas completar con la fecha `1932-01-01`. Mostrar las 10 primeras fechas.

- Normalizar los horas al formato de 24hs, las horas inválidas se procesaran como `00:00`. Mostrar las 10 primeras horas.

- Normalizar las ubicaciones. Quitar caracteres especiales y pasar todo a mayúsculas. Mostrar las 10 primeras ubicaciones.

- Limpiar y normalizar las patentes, en caso de error completar con pd.NA (no disponible/not available). Quitar caracteres especiales y pasar todo a mayúsculas. Mostrar las 10 últimas patentes.

- Eliminar las filas que posean valores relevantes para las multas y que se encuentren vacios. Mostrar la cantidad de registros eliminados y cuales son las columnas que se observaron utilizando la leyenda
```
Se eliminaron X filas.
Las columnas observadas son:
 - XXX
 - XXX
 - ...
```

- Detectar y eliminar los outliers. Mostrar la cantidad de registros eliminados y cuales son las columnas que se observaron utilizando la leyenda
```
Se eliminaron X filas.
Las columnas observadas son:
 - XXX
 - XXX
 - ...
```

- Crear la columna `exceso_velocidad_real` y calcularla. El exceso de velocidad real es la diferencia entre velocidad_registrada y la velocidad_maxima. Mostrar las primeras 10 patentes y su exceso de velocidad real.

- Crear la columna `exceso_velocidad` y calcularla. El exceso de velocidad es la diferencia entre velocidad_registrada y la velocidad_maxima más un 5 porciento de la velocidad máxima. Mostrar las primeras 10 patentes y su exceso de velocidad.

- Eliminar las filas que no cometieron infracciones según el exceso de velocidad.Mostrar cuantas filas se eliminaron con la leyenda
```
Se eliminaron X filas.
```

- Guardar el dataset limpio en `urban_flow/data/interim/speeding_fines.csv`

In [28]:
#@title Normalización de fechas del dataset

# Copia de trabajo para no perder el original
df_dates: pd.DataFrame = df.copy()

# Convertir a datetime con coerción de errores
df_dates["fecha"] = pd.to_datetime(
  df_dates["fecha"],
  errors="coerce"
)

# Reemplazar valores inválidos (NaT) con fecha fija
df_dates["fecha"] = df_dates["fecha"].fillna(
  pd.Timestamp("1932-01-01")
)

# Formatear a YYYY-MM-DD
df_dates["fecha"] = df_dates["fecha"].dt.strftime("%Y-%m-%d")

# Mostrar las primeras 10 fechas
df_dates["fecha"].head(10)




/tmp/ipykernel_5956/248608690.py:7: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_dates["fecha"] = pd.to_datetime(


,fecha
0,2021-09-24
1,2020-05-19
2,1932-01-01
3,1932-01-01
4,1932-01-01
5,2023-07-27
6,1932-01-01
7,1932-01-01
8,1932-01-01
9,1932-01-01


## Ejercicio 04

> Puntos: 2

Definir una clase FineAnalyzer que cumpla con las siguientes funcionalidades:

- Reciba el DataFrame limpio como argumento de inicialización.

- Encapsule correctamente los datos.

- Implemente métodos para:

  - Ranking de las patentes más multadas (top 5). Debe retornar un DataFrame ordenado de mayor a menor y el índice debe comenzar desde 1, con las columnas de patentes y cantidad..

  - Ranking de los horarios donde se producen más multas (top 5). Debe retornar un DataFrame ordenado de mayor a menor y el índice debe comenzar desde 1, con las columnas de hora y cantidad.


  - Exceso promedio de velocidad. Debe retorna un flotante.

  - Exceso real promedio de velocidad . Debe retorna un flotante.

  - Contabilizar la cantidad de multas por cada  ubicación. Debe retornar un DataFrame ordenado por ubicación mostrando la ubicación y la cantidad.

- Crear el objecto e invocar cada uno de los métodos en celdas separadas.
  

## Punto 05

> Puntos: 2


Responder a cada punto con un gráfico.

- Ranking de las 10 patentes más reincidentes ordenadas de mayor a menor. Exportar el gráfico con el nombre `ubran_flow/data/interim/plots/fines.jpg`


- Mostrar el porcentaje de infracciones por hora en un gráfico de torta. Exportar el gráfico con el nombre `ubran_flow/data/interim/plots/hours.jpg`

- Mostrar la cantidad de infracciones por mes en un gráfico de barras horizontal ordenado de mayor a menor. Exportar el gráfico con el nombre `ubran_flow/data/interim/plots/months.jpg`

- Mostrar en un gráfico de líneas de los excesos de velocidad agrupados por la hora 00:00. Exportar el gráfico con el nombre `ubran_flow/data/interim/plots/hour.jpg`

- Mostrar en un gráfico de líneas de los excesos de velocidad agrupados por la fecha 1932-01-01. Exportar el gráfico con el nombre `ubran_flow/data/interim/plots/date.jpg`

## Punto 06

> Puntos: 1

- Mostrar el porcentaje de infracción que se produjeron en la fecha 1932-01-01.

```
El porcentaje de infracciones en la fecha 1932-01-01 es XX.XX%
```

- Mostrar el porcentaje de infracción que se produjeron en la hora 00:00.
```
El porcentaje de infracciones a la hora 00:00 es XX.XX%
```

## Punto 07

> Puntos: 1



Redactar una conclusión acerca de los datos que contiene el dataset.


In [ ]:
%%writefile -a data/README.md
